client/client.cc

```diff
@@ -80,87 +87,273 @@ Status KvsClient::DeleteGrpc(const std::string& key,
-  Status status = client.DeleteGrpc(key, &response);
-  if (status.ok()) {
-    LOG(INFO) << "[Client]: Delete successful";
-    return 0;
-  }
+  Status status;
 
-  if (status.error_code() != grpc::StatusCode::PERMISSION_DENIED) {
-    LOG(INFO) << "[Client]: Error from KVS server: " << status.error_code();
-    return -1;
-  }
+  while (true) {
+    auto channel = grpc::CreateChannel(nodes[node_id].GetAddressPortStr(),
+                                       grpc::InsecureChannelCredentials());
+
+    std::unique_ptr<Kvs::Stub> stub = Kvs::NewStub(channel);
+
+    ClientContext context;
+    status = stub->Delete(&context, request, &response);
+    if (status.ok()) {
+      LOG(INFO) << response.status();
+      return 0;
+    }
+    if (status.error_code() == grpc::StatusCode::UNAVAILABLE) {
+      node_id = (++node_id) % nodes.size();
+      continue;
+    }
 
-  if (status.error_message().empty()) {
-    return -1;
+    if (status.error_code() != grpc::StatusCode::PERMISSION_DENIED) {
+      LOG(INFO) << "[Client]: Error from KVS server: " << status.error_code();
+      return -1;
+    }
+
+    if (status.error_message().empty()) {
+      LOG(INFO) << "[Client]: Error from KVS server***";
+      return -1;
+    }
+
+    // Retry this opeartion now with the leader address
+    auto ch = grpc::CreateChannel(status.error_message(),
+                                  grpc::InsecureChannelCredentials());
+
+    std::unique_ptr<Kvs::Stub> st = Kvs::NewStub(ch);
+
+    ClientContext ctx;
+    status = st->Delete(&ctx, request, &response);
+    if (status.ok()) {
+      return 0;
+    } else if (status.error_code() == grpc::StatusCode::UNAVAILABLE) {
+      absl::SleepFor(absl::Seconds(1));
+    } else {
+      LOG(WARNING) << "[Client]: Error from KVS server leader node: "
+                   << status.error_code();
+      return -1;
+    }
   }
 
-  // Retry this opeartion now with the leader address returned by the
-  KvsClient leader_client(grpc::CreateChannel(
-      status.error_message(), grpc::InsecureChannelCredentials()));
+  return -1;
+}
+
+int NonInteractiveClientTest(void) {
+  const std::vector<Node> nodes =
+      ParseNodesConfig(absl::GetFlag(FLAGS_kvs_node_config_file));
+
+  std::map<std::string, std::string> kv;
+
+  std::random_device rd;
+  std::mt19937 gen(rd());
+  std::uniform_int_distribution<> dis(1, 3);
+
+  constexpr size_t num_tests = 500;
+  for (size_t i = 0; i < num_tests; i++) {
+    int step = dis(gen);
+
+    switch (step) {
+      case 1: {
+        std::string k = "key" + std::to_string(i);
+        std::string v = "value" + std::to_string(i);
+        if (PutHelper(nodes, k, v) != 0) {
+          LOG(FATAL) << "[Client]: Put operation failed for test number: " << i;
+        } else {
+          LOG(INFO) << "[Client]: Put operation successful for: " << i;
+        }
+        kv[k] = v;
+        break;
+      }
+      case 2: {
+        std::uniform_int_distribution<> key_dis(0, static_cast<int>(i));
```

Comment:
do you just want the key to be in a range [0, i] on purpose?

Reply:
Yeah, I plan to follow this up with more tests.